### PR TITLE
Add ray intersection tests for primitive shapes

### PR DESCRIPTION
**Pull request by matt439**
# Description

There has been growing interest in using SplashKit to perform ray-tracing in both 2D and 3D environments. However, there is a lack of ray-based intersection functions with primitive shapes. I have added ray intersection functions for rectangles, circles, triangles, and quads.

Each shape has been given a function which can output the the intersection position and distance. I have also included an overload of each function that only returns a Boolean. By including the overload, the user is not required to pass arguments by reference which may be confusing to SIT102 students.

I collaborated with Shaun Ratcliff to complete the logic for the intersection functions.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

I added a new real-time graphical test to `sktest` in `test_geometry.cpp` to demonstrate the functions. It allows the user to move a ray's origin with the arrow keys, while the heading is determined by the mouse cursor's position. All four shapes are present for testing. If an intersection occurs, the corresponding shape is highlighted and the intersection point is shown.

## Testing Checklist

- [x] Tested with sktest

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: coresdk/src/coresdk/triangle_geometry.cpp [hunk 1]
```diff
@@ -102,6 +103,67 @@ namespace splashkit_lib
...
+        for (int i = 0; i < 3; ++i)
+        {
+            point_2d start_point = tri.points[i];                  // Start point of the edge
+            point_2d end_point = tri.points[(i + 1) % 3];          // End point of the edge
+
+            vector_2d edge_vector = vector_point_to_point(start_point, end_point); // Edge vector
+            vector_2d origin_to_edge = vector_point_to_point(origin, start_point);
+
+            // Cross product to determine parallelism
+            double cross_ray_and_edge = (heading.x * edge_vector.y) - (heading.y * edge_vector.x);
```
Comment (NoahJCross) [inline coresdk/src/coresdk/triangle_geometry.cpp:135]: The code doesn't check for zero vectors in the heading parameter. This could lead to division by zero in this calculation. could add something like this:
const double ZERO_MAGNITUDE_THRESHOLD = 1e-10;
if (vector_magnitude(heading) < ZERO_MAGNITUDE_THRESHOLD)

File: coresdk/src/test/test_geometry.cpp [hunk 2]
```diff
@@ -239,10 +248,133 @@ void test_triangle()
     close_window(w1);
 }
 
+void ray_shape_intersection(const line& l, const point_2d& circ_center, color& clr)
```
Comment (NoahJCross) [inline coresdk/src/test/test_geometry.cpp:251]: function name is slightly misleading as it's actually drawing visualization.

File: coresdk/src/test/test_geometry.cpp [hunk 3]
```diff
@@ -239,10 +248,133 @@ void test_triangle()
...
+    auto t1_color = COLOR_BLUE;
+    auto t1_hit_point = point_at(0.0, 0.0);
+    double t1_distance = 0.0;
+    bool t1_intersection = false;
+    auto q1 = quad_from(100.0, 300.0, 200.0, 350.0, 100.0, 550.0, 200.0, 500.0);
+    auto q1_color = COLOR_BLUE;
+    auto q1_hit_point = point_at(0.0, 0.0);
+    double q1_distance = 0.0;
+    bool q1_intersection = false;
+    auto player = point_at(300.0, 300.0);
```
Comment (NoahJCross) [inline coresdk/src/test/test_geometry.cpp:310]: Something like this would be a little cleaner and more clear:
struct TestShape {
    color clr;
    point_2d hit_point;
    double distance;
    bool intersection;
};

TestShape rect_data{COLOR_BLUE, point_at(0,0), 0.0, false};

File: coresdk/src/coresdk/triangle_geometry.cpp [hunk 4]
```diff
@@ -102,6 +103,67 @@ namespace splashkit_lib
...
+            if (std::fabs(cross_ray_and_edge) < 1e-8) continue; // Skip edges nearly parallel to the ray
```
Comment (NoahJCross) [inline coresdk/src/coresdk/triangle_geometry.cpp:136]: 1e-8 could be a const, PARALLEL_THRESHOLD seems like an apropriate name.

**Review by NoahJCross — CHANGES_REQUESTED** [review]
There are some structural and safety improvements that could be made, but the core functionality is good, with the ray intersection calculations working correctly.

File: coresdk/src/coresdk/triangle_geometry.cpp (hunk 1 above)
Comment (matt439) [reply coresdk/src/coresdk/triangle_geometry.cpp]: I added checks for a zero vector heading in the rectangle, circle, triangle and quad functions. I opted to use the __DBL_EPSILON__ constant as it is already defined and has more meaning than an arbitrary value.

File: coresdk/src/test/test_geometry.cpp (hunk 2 above)
Comment (matt439) [reply coresdk/src/test/test_geometry.cpp]: Changed name to `draw_ray_shape_intersection`.

File: coresdk/src/test/test_geometry.cpp (hunk 3 above)
Comment (matt439) [reply coresdk/src/test/test_geometry.cpp:306]: Added struct `test_shape`.

File: coresdk/src/coresdk/triangle_geometry.cpp (hunk 4 above)
Comment (matt439) [reply coresdk/src/coresdk/triangle_geometry.cpp]: Used __DBL_EPSILON__ again here.

**Comment by matt439** [comment]
I found another issue where if the heading is a not a unit vector, the functions are buggy. I altered the functions to convert the heading to a unit vector at the beginning of each function. The zero vector check is performed on the unit heading.

**Comment by NoahJCross** [comment]
> I found another issue where if the heading is a not a unit vector, the functions are buggy. I altered the functions to convert the heading to a unit vector at the beginning of each function. The zero vector check is performed on the unit heading.

That's a good catch. A non normalized vector would definitely cause issues with the hit_distance.

**Review by NoahJCross — APPROVED** [review]
I've reviewed this PR and approve of these changes.

The ray intersection tests add important features to SplashKit while keeping high standards:

# Code Quality
- Good use of doubles for accurate geometric calculations 
- Proper vector normalization for accurate distance measurements
- Good handling of edge cases (zero vectors, interior points)
- Clean code that fits well with existing geometry systems

# Testing
- Good test coverage in test_geometry.cpp

# Documentation
- Clear methods for both detailed and simple intersection checks
- Consistent design across different shapes
- Well-documented code that explains how it works

The changes add useful geometry features while keeping types consistent and calculations stable. 
Approved.

**Review by WhyPenguins — APPROVED** [review]
Looks good! :smiley: Have tested and it works well, code is clean and easy to read too.

I think this can be merged as-is, but just wanted to mention that it might have been simpler/better to write a `polygon_ray_intersection(const point_2d &origin, const vector_2d &heading, const vector<line> lines, point_2d &hit_point, double &hit_distance)` function based on the logic in the `triangle_ray_intersection`, which the quad, rectangle, and triangle functions all use internally.

I think if we add further shapes then this would be a good way to reduce the duplication :smiley:
